_drafts/GDB Tips and Tricks.md

```diff
@@ -0,0 +1,500 @@
+
+For now, all you need to do is download it, open the package and copy the application file to your Mac **Applications** folder so you could launch it in the later steps.
+
+#### 6. Adding Necessary Folders to $PATH
+In order to access the necessary commands that we'll be calling from anywhere in your system, you need to add their paths to the system **$PATH** environment variable.
+
+To do that, open up a Terminal session and type the following:
+
+```terminal
+$ sudo vi /etc/paths
+```
+
+You will be prompted to enter your administrator password. Once you've done that, go ahead and add the following lines to the bottom of the file:
+
+```
+<Your Folder>/nRF-Command-Line-Tools_9_8_1_OSX/nrfjprog
+<Your Folder>/gcc-arm-none-eabi-7-2017-q4-major/bin
+``` 
+
+Make sure to replace the `<Your Folder>` part with the appropriate folder name where you placed those packages on your system.
+
+We don't need to add any paths for the SEGGER J-Link software because it ran as an installer and would have already added the necessary binaries to our system.
+
+After you're done editing and saving the `paths` file, close the Terminal and then start a new session to make sure the updated `paths` file gets loaded.
+
+To verify, you can type the following command:
+
+```terminal
+$ echo $PATH
+/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/mafaneh/Memfault/nRF-Command-Line-Tools_9_8_1_OSX/nrfjprog:/Users/mafaneh/Memfault/gcc-arm-none-eabi-7-2017-q4-major/bin
+``` 
+
+That's it for the setup part, now on to the fun part!
+
+## Running the Example
+Before we can actually debug the application and use GDB, we need to build the example and flash it to the development kit.
+
+Following are the steps we need to take:
+### 1. nRF52 Example Compilation
+In our tutorial, we'll be using the UART example included as part of the nRF5 SDK (version 15.3.0). This example is located at:
+
+```
+<nRF5_SDK_Folder>/examples/peripheral/uart/
+```
+
+You'll notice there are many subfolders in that folder. We are mostly interested in the following highlighted folder in the screenshot:
+
+![](Graphics/uart_folder.png)
+<br/>
+
+Before we build the example, let's make sure we have the right compiler flags for debugging. This is necessary to include debugging symbols that help GDB present useful debugging information to the user. 
+
+To check the compiler flags, edit the `Makefile` located at `<nRF5_SDK_Folder>/examples/peripheral/uart/pca10056/blank/armgcc`.
+
+In the Makefile, look for the "common C flags" section:
+
+```bash
+# C flags common to all targets
+CFLAGS += $(OPT)
+CFLAGS += -DBOARD_PCA10056
+CFLAGS += -DBSP_DEFINES_ONLY
+CFLAGS += -DCONFIG_GPIO_AS_PINRESET
+CFLAGS += -DFLOAT_ABI_HARD
+CFLAGS += -DNRF52840_XXAA
+CFLAGS += -mcpu=cortex-m4
+CFLAGS += -mthumb -mabi=aapcs
+CFLAGS += -Wall -Werror
+```
+
+Let's modify this to reflect the following changes:
+
+```bash
+# C flags common to all targets
+#CFLAGS += $(OPT)
```

Comment:
I wouldn't completely disable optimizations, instead it's easiest to use `-Og` or even just `-O1` when you're debugging.

Worth adding a line to say all optimization levels will work, but some variables will disappear in higher optimization levels as they get optimized out.

Reply:
Got it. I'll make that change.